app/cmd/rpc/mesh/session.go

```diff
@@ -89,124 +87,494 @@ func (sn DispatchResponse) GetSupportedNodes() []string {
 	return nodes
 }
 
-type AppSessionCache struct {
-	PublicKey       string
-	Chain           string
-	Dispatch        *DispatchResponse
+// NodeSession - contains error/valid information for the node-session relation
+type NodeSession struct {
+	PubKey string
+	// todo: if we can figure out a way to check this, otherwise receive/notify relays until fullNode return evidence sealed.
+	// ^^ that could work with problem, just few "free relays"
 	RemainingRelays int64
+	RelayMeta       *pocketTypes.RelayMeta
+	Validated       bool
+	RetryTimes      int
 	IsValid         bool
 	Error           *SdkErrorResponse
+	Session         *Session
 }
 
-// getAppSession - call ServicerURL to get an application session using retrieve header
-func getAppSession(relay *pocketTypes.Relay, model interface{}) *SdkErrorResponse {
-	servicerNode := getServicerFromPubKey(relay.Proof.ServicerPubKey)
+func (ns *NodeSession) CountRelay() (*NodeSession, bool) {
+	if !ns.Validated {
+		// if this session is not validated yet, will keep been optimistic
+		return ns, true
+	}
+
+	ns.RemainingRelays -= 1
+
+	if ns.RemainingRelays > 0 {
+		return ns, true // still can send relays
+	}
+
+	address, _ := GetAddressFromPubKeyAsString(ns.PubKey)
+
+	logger.Debug(
+		fmt.Sprintf(
+			"servicer=%s exhaust relays for app=%s chain=%s sessionHeight=%d",
+			address,
+			ns.Session.AppPublicKey,
+			ns.Session.Chain,
+			ns.Session.BlockHeight,
+		),
+	)
+
+	ns.IsValid = false
+	ns.Error = NewSdkErrorFromPocketSdkError(pocketTypes.NewOverServiceError(ModuleName))
+
+	return ns, false
+}
+
+// Session - Contains general app session information
+type Session struct {
+	Hash         string
+	AppPublicKey string
+	Chain        string
+	BlockHeight  int64
+	Dispatch     *DispatchResponse
+	Nodes        *xsync.MapOf[string, *NodeSession]
+}
+
+func (ns *NodeSession) ReScheduleValidationTask(session *Session, servicerPubKey string) {
+	if ns.RetryTimes > 20 {
```

Comment:
Set 20 as a const to avoid magic numbers
https://en.wikipedia.org/wiki/Magic_number_(programming)

Reply:
thanks will fix, good catch

Reply:
I set this to a default value. `SessionStorageValidateRetryMaxTimes`